lib/config.ml

```diff
@@ -1,24 +1,95 @@
-module Formats = struct
+open Prelude
 
-  type htransform = string -> string
-  type dtransform = string -> string
-  
+module Formats = struct
   type variety =
-    | DataOnly of dtransform
-    | DataAndHeader of (htransform * dtransform)
+    | DataOnly
+    | DataAndHeader
     | NoChange
 
+  type transform_data =
+    { target_type: string;
+      shell_command: string;
+      variety: variety
+    }
+
   module Dict = Map.Make (String)
-    
-  type t = variety Dict.t
+
+  type t = (transform_data list) Dict.t
 
   module Error = struct
     type t =
       | ReferParse of string
-      | Unix of (string * Unix.error)
-      | MimeParse of string
-      | CharacterEncoding of string
+      | ConfigData of string
+
+    let map_msg f err =
+      match err with
+      | ReferParse msg -> ReferParse (f msg)
+      | ConfigData msg -> ConfigData (f msg)
   end
+
   type error = Error.t
+end
+
+module ParseConfig = struct
+  open Formats
+
+  type config_entry =
+    { source_type   : string ;
+      target_type   : string ;
+      shell_command : string ;
+    }
+
+  let entry_of_assoc config_assoc =
+    let construct_config_entry st tt ss =
+      { source_type   = st ;
+        target_type   = tt ;
+        shell_command = ss ;
+      }
+    in
+    let check key err_msg =
+      Option.to_result
+        ~none:(Error.ConfigData ("Config File Error: " ^ err_msg))
+        (assoc_opt key config_assoc)
+    in
+    let  ( let* ) = Result.(>>=)                                 in
+    let* st       = check "source_type"   "no source type given" in
+    let* tt       = check "target_type"   "no target type given" in
+    let* ss       = check "shell_command" "no script path given" in
+    Ok (construct_config_entry st tt ss)
+
+  let transform_data_of_entry entry =
+    let variety_of_entry =
+      if   entry.source_type = entry.target_type
+      then DataOnly
+      else DataAndHeader
+    in
+    { target_type   = entry.target_type   ;
+      shell_command = entry.shell_command ;
+      variety       = variety_of_entry    ;
+    }
+
+  let parse_config_str config_str =
```

Comment:
thanks, though they could definitely be improved. I think in the future I want to design the Error constructors to take more metadata (line number, relevant line, etc.) so that we have more freedom in constructing the messages and so we can factor out message construction from the parsing process.

Reply:
Yes to all that!